File: go/resolver.go
package octoauth

import (
	"bytes"
	"context"
	"encoding/hex"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"strings"
	"time"
)

const resolveEndpoint = "/v1/internal/auth/resolve"

// BotResolveMode selects Bot-self identity or owner delegation.
type BotResolveMode string

const (
	ModeAsBot BotResolveMode = "AS_BOT"
	ModeOBO   BotResolveMode = "OBO"
)

type ResolveResource struct {
	Type string `json:"type"`
	ID   string `json:"id"`
}

type ResolveRequest struct {
	Mode     BotResolveMode   `json:"mode"`
	SpaceID  string           `json:"space_id"`
	Action   string           `json:"action,omitempty"`
	Resource *ResolveResource `json:"resource,omitempty"`
}

type ResolvedIdentity struct {
	UID     string `json:"uid"`
	Kind    string `json:"kind"`
	SpaceID string `json:"space_id"`
}

type ResolvedDelegation struct {
	GrantID       int64  `json:"grant_id"`
	MatchedScope  string `json:"matched_scope"`
	PolicyVersion int64  `json:"policy_version"`
	Action        string `json:"action"`
	DecisionID    string `json:"decision_id"`
}

type ResolvedPrincipal struct {
	Mode       BotResolveMode      `json:"mode"`
	Actor      ResolvedIdentity    `json:"actor"`
	Subject    ResolvedIdentity    `json:"subject"`
	Delegation *ResolvedDelegation `json:"delegation,omitempty"`
}

type Resolver interface {
	Resolve(ctx context.Context, credential string, req ResolveRequest) (*ResolvedPrincipal, error)
}

type botResolver struct{ cfg *Config }

// NewResolver constructs the Bot resolver using the same Bot credential as
// the legacy verifier. The caller must derive Action from a controlled route.
func NewResolver(cfg *Config) Resolver {
	applyDefaults(cfg)
	if strings.TrimSpace(cfg.BaseURL) == "" {
		panic("octoauth: Resolver requires BaseURL")
	}
	return &botResolver{cfg: cfg}
}

func invalidResolveRequest(message string) error {
	return &Error{Kind: ErrKindInvalidRequest, Message: message, Verifier: KindBot}
}

func resolveProtocolFailure(message string, cause error) error {
	return &Error{Kind: ErrKindInfraFailure, Message: message, Cause: cause, Verifier: KindBot}
}

func (r *botResolver) Resolve(ctx context.Context, credential string, request ResolveRequest) (principal *ResolvedPrincipal, err error) {
	started := time.Now()
	defer func() {
		result := verifyResultMissOK
		if err != nil {
			result = verifyResultInfra
			var authErr *Error
			if errors.As(err, &authErr) {
				r.cfg.Metrics.IncErrorByKind(KindBot, authErr.Kind)
				if authErr.Kind != ErrKindInfraFailure {
					result = verifyResultAuthErr
				}
			}
		}
		r.cfg.Metrics.ObserveVerifyDuration(KindBot, result, time.Since(started))
	}()
	if !strings.HasPrefix(credential, PrefixBF) || len(credential) <= len(PrefixBF) || strings.TrimSpace(credential) != credential {
		return nil, &Error{Kind: ErrKindInvalidCredential, Message: "unsupported or empty Bot credential", Verifier: KindBot}
	}
	if strings.TrimSpace(request.SpaceID) == "" {
		return nil, invalidResolveRequest("space_id is required")
	}
	switch request.Mode {
	case ModeAsBot:
		if request.Action != "" || request.Resource != nil {
			return nil, invalidResolveRequest("AS_BOT cannot carry Action or resource")
		}
	case ModeOBO:
		if strings.TrimSpace(request.Action) == "" {
			return nil, invalidResolveRequest("OBO requires Action")
		}
	default:
		return nil, invalidResolveRequest("invalid Bot resolve mode")
	}

	input := struct {
		BotToken string `json:"bot_token"`
		ResolveRequest
	}{BotToken: credential, ResolveRequest: request}
	payload, err := json.Marshal(input)
	if err != nil {
		return nil, resolveProtocolFailure("encode Resolve request", err)
	}
	httpReq, err := http.NewRequestWithContext(ctx, http.MethodPost, r.cfg.BaseURL+resolveEndpoint, bytes.NewReader(payload))
	if err != nil {
		return nil, resolveProtocolFailure("build Resolve request", err)
	}
	httpReq.Header.Set("Content-Type", "application/json")
	httpReq.Header.Set("Accept", "application/json")
	httpReq.Header.Set("Cache-Control", "no-store")
	resp, err := r.cfg.HTTPClient.Do(httpReq)
	if err != nil {
		return nil, resolveProtocolFailure("Resolve request failed", err)
	}
	defer resp.Body.Close()
	body, err := io.ReadAll(io.LimitReader(resp.Body, maxVerifyBodyBytes+1))
	if err != nil || len(body) > maxVerifyBodyBytes {
		return nil, resolveProtocolFailure("read bounded Resolve response", err)
	}
	if resp.StatusCode != http.StatusOK {
		logVerifyResult(r.cfg, KindBot, resolveEndpoint, resp.StatusCode, body)
		return nil, resolveWireError(resp.StatusCode, body)
	}
	var out ResolvedPrincipal
	if err := json.Unmarshal(body, &out); err != nil {
		return nil, resolveProtocolFailure("decode Resolve response", err)
	}
	if out.Mode != request.Mode || out.Actor.UID == "" || out.Actor.Kind != "BOT" ||
		out.Actor.SpaceID != request.SpaceID || out.Subject.UID == "" || out.Subject.SpaceID != request.SpaceID {
		return nil, resolveProtocolFailure("Resolve response identity mismatch", nil)
	}
	if request.Mode == ModeAsBot {
		if out.Subject.Kind != "BOT" || out.Subject.UID != out.Actor.UID || out.Delegation != nil {
			return nil, resolveProtocolFailure("AS_BOT response mismatch", nil)
		}
	} else if out.Subject.Kind != "HUMAN" || out.Delegation == nil || out.Delegation.GrantID <= 0 ||
		out.Delegation.PolicyVersion <= 0 || !validResolveDecisionID(out.Delegation.DecisionID) ||
		out.Delegation.Action != request.Action || out.Delegation.MatchedScope != "ALL" {
		return nil, resolveProtocolFailure("OBO response mismatch", nil)
	}
	return &out, nil
}

func resolveWireError(status int, body []byte) error {
	var wire struct {
		Code      string `json:"code"`
		RequestID string `json:"request_id"`
	}
	if err := json.Unmarshal(body, &wire); err != nil {
		return resolveProtocolFailure("invalid Resolve error envelope", err)
	}
	if wire.RequestID == "" {
		return resolveProtocolFailure("Resolve error response missing request_id", nil)
	}
	kind := ErrKindInfraFailure
	valid := false
	switch wire.Code {
	case "invalid_mode", "invalid_request", "missing_space_id":
		kind, valid = ErrKindInvalidRequest, status == http.StatusBadRequest
	case "invalid_credential":
		kind, valid = ErrKindInvalidCredential, status == http.StatusUnauthorized
	case "disabled":
		kind, valid = ErrKindDisabled, status == http.StatusForbidden
	case "space_not_allowed", "delegation_denied", "action_not_allowed":
		kind, valid = ErrKindForbidden, status == http.StatusForbidden
	case "infra_failure":
		kind, valid = ErrKindInfraFailure, status == http.StatusServiceUnavailable
	}
	if !valid {
		return resolveProtocolFailure(fmt.Sprintf("unexpected Resolve status/code: %d/%s", status, wire.Code), nil)
	}
	return &Error{Kind: kind, Code: wire.Code, RequestID: wire.RequestID, Message: "Resolve rejected", Verifier: KindBot}
}

func validResolveDecisionID(value string) bool {
	if len(value) != 32 || value != strings.ToLower(value) {
		return false
	}
	_, err := hex.DecodeString(value)
	return err == nil
}

File: ts/src/resolver.ts
/** New Bot-only identity/delegation resolver. Legacy verifiers are unchanged. */
import type { Config } from './config.js'
import { applyDefaults } from './config.js'
import { OctoAuthError } from './errors.js'
import {
  logVerifyResult,
  readTextBounded,
  VERIFY_RESULT_AUTH_ERR,
  VERIFY_RESULT_INFRA,
  VERIFY_RESULT_MISS_OK,
} from './http-verifier.js'
import { KindBot, PrefixBF } from './verifier.js'

const ENDPOINT = '/v1/internal/auth/resolve'
const MAX_BODY_BYTES = 1 << 20
const DECISION_ID_PATTERN = /^[0-9a-f]{32}$/

export type BotResolveMode = 'AS_BOT' | 'OBO'

export interface ResolveRequest {
  mode: BotResolveMode
  spaceId: string
  action?: string
  resource?: { type: string; id: string }
}

export interface ResolvedIdentity {
  uid: string
  kind: 'BOT' | 'HUMAN'
  spaceId: string
}

export interface ResolvedDelegation {
  grantId: number
  matchedScope: 'ALL'
  policyVersion: number
  action: string
  decisionId: string
}

export interface ResolvedPrincipal {
  mode: BotResolveMode
  actor: ResolvedIdentity
  subject: ResolvedIdentity
  delegation?: ResolvedDelegation
}

export interface BotResolver {
  resolve(credential: string, request: ResolveRequest, signal?: AbortSignal): Promise<ResolvedPrincipal>
}

interface WireIdentity { uid?: string; kind?: string; space_id?: string }
interface WireDelegation {
  grant_id?: number
  matched_scope?: string
  policy_version?: number
  action?: string
  decision_id?: string
}
interface WirePrincipal {
  mode?: string
  actor?: WireIdentity
  subject?: WireIdentity
  delegation?: WireDelegation
}

function requestError(message: string): OctoAuthError {
  return new OctoAuthError('invalid-request', message, { verifier: KindBot })
}

function protocolError(message: string, cause?: unknown): OctoAuthError {
  return new OctoAuthError('infra-failure', message, { verifier: KindBot, cause })
}

function wireError(status: number, text: string): OctoAuthError {
  let envelope: { code?: string; request_id?: string }
  try { envelope = JSON.parse(text) as typeof envelope } catch (err) {
    return protocolError('invalid Resolve error envelope', err)
  }
  if (typeof envelope?.request_id !== 'string' || envelope.request_id.length === 0) {
    return protocolError('Resolve error response missing request_id')
  }
  const code = envelope?.code
  let kind: 'invalid-request' | 'invalid-credential' | 'disabled' | 'forbidden' | 'infra-failure' = 'infra-failure'
  let valid = false
  switch (code) {
    case 'invalid_mode':
    case 'invalid_request':
    case 'missing_space_id':
      kind = 'invalid-request'; valid = status === 400; break
    case 'invalid_credential':
      kind = 'invalid-credential'; valid = status === 401; break
    case 'disabled':
      kind = 'disabled'; valid = status === 403; break
    case 'space_not_allowed':
    case 'delegation_denied':
    case 'action_not_allowed':
      kind = 'forbidden'; valid = status === 403; break
    case 'infra_failure':
      kind = 'infra-failure'; valid = status === 503; break
  }
  if (!valid) return protocolError(`unexpected Resolve status/code: ${status}/${code ?? ''}`)
  return new OctoAuthError(kind, 'Resolve rejected', {
    verifier: KindBot,
    code,
    requestId: envelope.request_id,
  })
}

function readIdentity(value: WireIdentity | undefined, spaceId: string): ResolvedIdentity | undefined {
  if (!value?.uid || value.space_id !== spaceId || (value.kind !== 'BOT' && value.kind !== 'HUMAN')) {
    return undefined
  }
  return { uid: value.uid, kind: value.kind, spaceId: value.space_id }
}

export function newBotResolver(cfg: Config): BotResolver {
  const resolved = applyDefaults(cfg)

  return {
    async resolve(credential, request, signal) {
      const started = Date.now()
      let result = VERIFY_RESULT_MISS_OK
      try {
      if (!credential.startsWith(PrefixBF) || credential.length <= PrefixBF.length || credential.trim() !== credential) {
        throw new OctoAuthError('invalid-credential', 'unsupported or empty Bot credential', { verifier: KindBot })
      }
      if (!request.spaceId?.trim()) throw requestError('spaceId is required')
      if (request.mode === 'AS_BOT') {
        if ((request.action !== undefined && request.action !== '') || request.resource !== undefined) {
          throw requestError('AS_BOT cannot carry action or resource')
        }
      } else if (request.mode === 'OBO') {
        if (!request.action?.trim()) throw requestError('OBO requires action')
      } else {
        throw requestError('invalid Bot resolve mode')
      }

      const timeout = AbortSignal.timeout(resolved.timeoutMs)
      const combined = signal ? AbortSignal.any([signal, timeout]) : timeout
      let response: Response
      try {
        response = await resolved.fetch(resolved.baseUrl + ENDPOINT, {
          method: 'POST', redirect: 'error', signal: combined,
          headers: {
            'Content-Type': 'application/json', Accept: 'application/json',
            'Cache-Control': 'no-store',
          },
          body: JSON.stringify({
            bot_token: credential, mode: request.mode, space_id: request.spaceId,
            ...(request.action ? { action: request.action } : {}),
            ...(request.resource === undefined ? {} : { resource: request.resource }),
          }),
        })
      } catch (err) {
        throw protocolError('Resolve request failed', err)
      }
      let text: string
      try { text = await readTextBounded(response, MAX_BODY_BYTES) } catch (err) {
        throw protocolError('read bounded Resolve response', err)
      }
      if (response.status !== 200) {
        logVerifyResult(resolved, KindBot, ENDPOINT, response.status, text)
        throw wireError(response.status, text)
      }
      let wire: WirePrincipal
      try { wire = JSON.parse(text) as WirePrincipal } catch (err) {
        throw protocolError('decode Resolve response', err)
      }
      const actor = readIdentity(wire?.actor, request.spaceId)
      const subject = readIdentity(wire?.subject, request.spaceId)
      if (!actor || !subject || actor.kind !== 'BOT' || wire.mode !== request.mode) {
        throw protocolError('Resolve response identity mismatch')
      }
      if (request.mode === 'AS_BOT') {
        if (subject.kind !== 'BOT' || subject.uid !== actor.uid || wire.delegation !== undefined) {
          throw protocolError('AS_BOT response mismatch')
        }
        return { mode: 'AS_BOT', actor, subject }
      }
      const delegation = wire.delegation
      if (subject.kind !== 'HUMAN' || !delegation || !Number.isSafeInteger(delegation.grant_id) ||
          delegation.grant_id! <= 0 || delegation.matched_scope !== 'ALL' ||
          delegation.action !== request.action || !Number.isSafeInteger(delegation.policy_version) ||
          delegation.policy_version! <= 0 || typeof delegation.decision_id !== 'string' ||
          !DECISION_ID_PATTERN.test(delegation.decision_id)) {
        throw protocolError('OBO response mismatch')
      }
      return {
        mode: 'OBO', actor, subject,
        delegation: {
          grantId: delegation.grant_id!, matchedScope: 'ALL',
          policyVersion: delegation.policy_version!, action: delegation.action!,
          decisionId: delegation.decision_id,
        },
      }
      } catch (err) {
        result = VERIFY_RESULT_INFRA
        if (err instanceof OctoAuthError) {
          resolved.metrics.incErrorByKind(KindBot, err.kind)
          if (err.kind !== 'infra-failure') result = VERIFY_RESULT_AUTH_ERR
        }
        throw err
      } finally {
        resolved.metrics.observeVerifyDuration(KindBot, result, Date.now() - started)
      }
    },
  }
}
